File: jar/src/main/java/jline/api/mam/m3pp/M3pp_rand.java
/**
 * @file M3PP random process generator
 *
 * @since LINE 3.0
 */
package jline.api.mam.m3pp;

import java.util.Random;

import jline.util.matrix.Matrix;
import jline.util.matrix.MatrixCell;

public final class M3pp_rand {
    private M3pp_rand() {}

    public static MatrixCell m3pp_rand(int order, int classes) {
        return m3pp_rand(order, classes, null);
    }

    /**
     * Generates a random M3PP with specified order and number of classes.
     */
    public static MatrixCell m3pp_rand(int order, int classes, Long seed) {
        Random random = (seed != null) ? new Random(seed) : new Random();

        MatrixCell mmpp = mmpp_rand(order, random);

        MatrixCell mmap = new MatrixCell(classes + 2);
        mmap.set(0, mmpp.get(0));
        mmap.set(1, mmpp.get(1));

        for (int i = 0; i < order; i++) {
            double[] p = new double[classes];
            double pSum = 0.0;
            for (int j = 0; j < classes; j++) {
                p[j] = random.nextDouble();
                pSum += p[j];
            }

            for (int j = 0; j < p.length; j++) {
                p[j] /= pSum;
            }

            for (int c = 0; c < classes; c++) {
                if (mmap.get(2 + c) == null) {
                    mmap.set(2 + c, new Matrix(order, order));
                }

                Matrix classMatrix = mmap.get(2 + c);
                Matrix baseMatrix = mmpp.get(1);

                for (int row = 0; row < order; row++) {
                    for (int col = 0; col < order; col++) {
                        classMatrix.set(row, col, baseMatrix.get(row, col) * p[c]);
                    }
                }
            }
        }

        return mmap;
    }

    private static MatrixCell mmpp_rand(int order, Random random) {
        Matrix D0 = new Matrix(order, order);
        Matrix D1 = new Matrix(order, order);

        for (int i = 0; i < order; i++) {
            for (int j = 0; j < order; j++) {
                if (i != j) {
                    D0.set(i, j, random.nextDouble() * 2.0);
                }
            }
        }

        for (int i = 0; i < order; i++) {
            D1.set(i, i, random.nextDouble() * 5.0 + 0.1);
        }

        // Normalize so that D0's diagonal makes (D0 + D1) a proper generator;
        // matches MATLAB mmpp_rand, which ends with map_normalize. Without it
        // the returned process is not a valid MAP (positive row sums).
        return jline.api.mam.Map_normalize.map_normalize(D0, D1);
    }

    public static MatrixCell m3pp_rand_targeted(int order, int classes) {
        return m3pp_rand_targeted(order, classes, 1.0, 2.0, null);
    }

    public static MatrixCell m3pp_rand_targeted(int order, int classes, double targetRate,
                                                double targetSCV, Long seed) {
        Random random = (seed != null) ? new Random(seed) : new Random();
        MatrixCell bestMmap = null;
        double bestError = Double.MAX_VALUE;

        for (int iter = 0; iter < 20; iter++) {
            MatrixCell candidate = m3pp_rand(order, classes, random.nextLong());

            double candidateRate = computeArrivalRate(candidate);
            double candidateSCV = computeSCV(candidate);

            double rateError = Math.abs(candidateRate - targetRate) / targetRate;
            double scvError = Math.abs(candidateSCV - targetSCV) / targetSCV;
            double totalError = rateError + scvError;

            if (totalError < bestError) {
                bestError = totalError;
                bestMmap = candidate;
            }
        }

        return bestMmap != null ? bestMmap : m3pp_rand(order, classes, seed);
    }

    private static double computeArrivalRate(MatrixCell mmap) {
        Matrix D1 = mmap.get(1);
        return D1.elementSum();
    }

    private static double computeSCV(MatrixCell mmap) {
        return 1.5;
    }
}

File: jar/src/main/kotlin/jline/api/mam/m3pp/M3pp_rand.kt
/**
 * @file M3PP random process generator
 * 
 * Implements random generation of Markovian Multi-class Point Processes (M3PP) 
 * with specified order and class count. Provides stochastic model creation 
 * for testing and simulation purposes with configurable statistical properties.
 * 
 * @since LINE 3.0
 */
package jline.api.mam.m3pp

import jline.util.matrix.MatrixCell
import jline.util.matrix.Matrix
import kotlin.random.Random

/**
 * Generates a random M3PP (Markovian Multi-class Point Process) with specified order and number of classes.
 *
 * @param order Order of the underlying MMPP
 * @param classes Number of classes
 * @param seed Random seed for reproducibility (optional)
 * @return Random M3PP as MatrixCell
 */
fun m3pp_rand(order: Int, classes: Int, seed: Long? = null): MatrixCell {
    val random = if (seed != null) Random(seed) else Random.Default
    
    // Generate random MMPP base
    val mmpp = mmpp_rand(order, random)
    
    // Create M3PP structure
    val mmap = MatrixCell(classes + 2)
    mmap[0] = mmpp[0] // D0 matrix
    mmap[1] = mmpp[1] // D1 matrix (total arrivals)
    
    // Distribute the arrival intensities across classes
    for (i in 0 until order) {
        // Generate random class probabilities for each state
        val p = DoubleArray(classes) { random.nextDouble() }
        val pSum = p.sum()
        
        // Normalize probabilities
        for (j in p.indices) {
            p[j] /= pSum
        }
        
        // Create class-specific matrices
        for (c in 0 until classes) {
            if (mmap[2 + c] == null) {
                mmap[2 + c] = Matrix(order, order)
            }
            
            // Scale the base arrival matrix by class probability
            val classMatrix = mmap[2 + c]
            val baseMatrix = mmpp[1] // D1 matrix
            
            for (row in 0 until order) {
                for (col in 0 until order) {
                    classMatrix[row, col] = baseMatrix[row, col] * p[c]
                }
            }
        }
    }
    
    return mmap
}

/**
 * Generates a random MMPP (Markovian Modulated Poisson Process) of given order.
 *
 * @param order Order of the MMPP
 * @param random Random number generator
 * @return Random MMPP as MatrixCell with D0 and D1 matrices
 */
private fun mmpp_rand(order: Int, random: Random): MatrixCell {
    val mmpp = MatrixCell(2)
    
    // Generate D0 matrix (generator matrix)
    val D0 = Matrix(order, order)
    val D1 = Matrix(order, order)
    
    // Fill off-diagonal elements of D0 with random positive values
    for (i in 0 until order) {
        var rowSum = 0.0
        
        for (j in 0 until order) {
            if (i != j) {
                val value = random.nextDouble() * 2.0 // Random transition rate
                D0[i, j] = value
                rowSum += value
            }
        }
        
        // Set diagonal to make rows sum to zero (generator property)
        D0[i, i] = -rowSum
    }
    
    // Generate D1 matrix (arrival rates)
    for (i in 0 until order) {
        for (j in 0 until order) {
            if (i == j) {
                // Diagonal elements: arrival rates in each state
                D1[i, j] = random.nextDouble() * 5.0 + 0.1 // Rate between 0.1 and 5.1
            } else {
                // Off-diagonal elements typically zero for MMPP
                D1[i, j] = 0.0
            }
        }
    }
    
    mmpp[0] = D0
    mmpp[1] = D1
    
    return mmpp
}

/**
 * Generates a random M3PP with specific characteristics.
 *
 * @param order Order of the underlying process
 * @param classes Number of classes  
 * @param targetRate Target overall arrival rate
 * @param targetSCV Target squared coefficient of variation
 * @param seed Random seed
 * @return Random M3PP with approximately target characteristics
 */
fun m3pp_rand_targeted(
    order: Int, 
    classes: Int, 
    targetRate: Double = 1.0,
    targetSCV: Double = 2.0,
    seed: Long? = null
): MatrixCell {
    val random = if (seed != null) Random(seed) else Random.Default
    var bestMmap: MatrixCell? = null
    var bestError = Double.MAX_VALUE
    
    // Generate multiple candidates and pick the best one
    repeat(20) {
        val candidate = m3pp_rand(order, classes, random.nextLong())
        
        // Compute characteristics (simplified)
        val candidateRate = computeArrivalRate(candidate)
        val candidateSCV = computeSCV(candidate)
        
        val rateError = Math.abs(candidateRate - targetRate) / targetRate
        val scvError = Math.abs(candidateSCV - targetSCV) / targetSCV
        val totalError = rateError + scvError
        
        if (totalError < bestError) {
            bestError = totalError
            bestMmap = candidate
        }
    }
    
    return bestMmap ?: m3pp_rand(order, classes, seed)
}

/**
 * Helper functions for computing basic characteristics
 */
private fun computeArrivalRate(mmap: MatrixCell): Double {
    // Simplified computation - would need full steady-state analysis
    val D1 = mmap[1]
    return D1.elementSum() // Rough approximation
}

private fun computeSCV(mmap: MatrixCell): Double {
    // Simplified computation - would need full moment analysis  
    return 1.5 // Placeholder
}
/**
 * M3Pp Rand algorithms
 */
@Suppress("unused")
class M3ppRandAlgo {
    companion object {
        // Class documentation marker for Dokka
    }
}